Src/IronPython/Runtime/StringFormatter.cs

```diff
@@ -859,10 +860,12 @@ internal static byte[] FormatBytes(CodeContext/*!*/ context, ReadOnlySpan<byte>
             Debug.Assert(_asBytes);
             if (_opts.Value is Bytes bytes || Bytes.TryInvokeBytesOperator(_context, _opts.Value, out bytes!)) {
                 AppendString(StringOps.Latin1Encoding.GetString(bytes.UnsafeByteArray));
-            } else if (_opts.Value is ByteArray byteArray) {
-                AppendString(StringOps.Latin1Encoding.GetString(byteArray.UnsafeByteList.AsByteSpan()));
+            } else if (_opts.Value is IBufferProtocol bufferProtocol) {
+                using var buffer = bufferProtocol.GetBuffer(BufferFlags.FullRO);
```

Comment:
Sorry, I've just noticed this. On it.

Reply:
Yes, `FullRO` is the proper flag. It mimics the constructor of `Bytes`, which also uses this flag. Also CPython uses this flag in `_PyBytes_FromBuffer`.

More generally, the flags indicate what kind of information/content/structure of the buffer the requesting party is capable and prepared to handle. So `Full` means that any type of buffer is OK and the burden is on the requesting party. The `Writable`/`RO` flags are a bit counter-intuitive, because `Writable` (bit 0) actually adds a constraint (rather than relaxes) meaning the buffer must be writable. Without this flag, the buffer may be either writable or readonly. Hence `FullRO` means in practice "no constraints whatsoever" and must always succeed. On the other hand, sole `Writable` is the most constrained: unformatted (i.e. byte-oriented), writable, contiguous blob with no structure.

> As a side note. I had inadvertently used `buffer.AsReadOnlySpan()` with the sliced memoryview which obviously produced incorrect results - should we maybe be throwing an exception from `AsReadOnlySpan`?

That would defeat the purpose of the flags. If you requested a buffer with `FullRO` (or any flag that included `Strides` (bit 8)), you declared that the span you get will be interpreted for the internal structure, even if the buffer is not contiguous. If an exception were to be thrown, it would make it impossible to access the buffer unless it happened to be contiguous. The buffer protocol is rather low level so incorrect use may lead to incorrect results without warning (like in _ctypes_). 

On my side note: this may not be worth the effort, but it is possible to avoid allocating the extra array in case the buffer is not C-contiguous if you used `buffer.EnumerateBytes()` and `StringExtensions.Create`. The question is how big those buffers can get in practice?

Reply:
> Yes, `FullRO` is the proper flag. It mimics the constructor of `Bytes`, which also uses this flag. Also CPython uses this flag in `_PyBytes_FromBuffer`.

Thanks for checking!

> > As a side note. I had inadvertently used `buffer.AsReadOnlySpan()` with the sliced memoryview which obviously produced incorrect results - should we maybe be throwing an exception from `AsReadOnlySpan`?
> 
> That would defeat the purpose of the flags. If you requested a buffer with `FullRO` (or any flag that included `Strides` (bit 8)), you declared that the span you get will be interpreted for the internal structure, even if the buffer is not contiguous. If an exception were to be thrown, it would make it impossible to access the buffer unless it happened to be contiguous. The buffer protocol is rather low level so incorrect use may lead to incorrect results without warning (like in _ctypes_).

Ok yeah, that makes sense - guess we just have to be careful.

> On my side note: this may not be worth the effort, but it is possible to avoid allocating the extra array in case the buffer is not C-contiguous if you used `buffer.EnumerateBytes()` and `StringExtensions.Create`. The question is how big those buffers can get in practice?

Probably not worth the effort - don't know that this code path will ever be used in practice. I just happened to fall on it because I was trying to guess what sort of buffer CPython allowed.